consumer.go

```diff
@@ -461,9 +461,47 @@ func (c *Consumer) Fetch(batchSize int) ([]*Msg, error) {
 		c.dlsMsgsMutex.Unlock()
 		return msgs, nil
 	}
+
+	c.conn.prefetchedMsgs.lock.Lock()
+	if prefetchedMsgsForStation, ok := c.conn.prefetchedMsgs.msgs[c.stationName]; ok {
+		if prefetchedMsgsForCG, ok := prefetchedMsgsForStation[c.Name]; ok {
+			if len(prefetchedMsgsForCG) > 0 {
+				if len(prefetchedMsgsForCG) <= batchSize {
+					msgs = prefetchedMsgsForCG
+					prefetchedMsgsForCG = []*Msg{}
+				} else {
+					msgs = prefetchedMsgsForCG[:batchSize-1]
+					prefetchedMsgsForCG = prefetchedMsgsForCG[batchSize-1:]
+				}
+				c.conn.prefetchedMsgs.msgs[c.stationName][c.Name] = prefetchedMsgsForCG
+			}
+		}
+	}
+	c.conn.prefetchedMsgs.lock.Unlock()
+	if prefetch {
+		go c.prefetchMsgs()
+	}
+	if len(msgs) > 0 {
+		return msgs, nil
+	}
 	return c.fetchSubscriprionWithTimeout()
 }
 
+func (c *Consumer) prefetchMsgs() {
+	c.conn.prefetchedMsgs.lock.Lock()
```

Comment:
just to make sure, the prefetch cache will always contain not more than 1 batch ahead?

Reply:
Yes, this is true.